docs/design/sql/14-stream-to-stream-join.md

```diff
@@ -0,0 +1,232 @@
+goals are:
+
+- SQL based / Non-Java friendly: easy to use format also for non-Java developers to use stream to stream joins
+- different join types (`INNER` , `LEFT`/`RIGHT`/`FULL` `OUTER` `JOIN`s.
+
+It should be possible to efficiently manage state for these joins using watermarks or other mechanism(s).
+
+### Functional Design
+
+#### Summary of Functionality
+
+User may execute stream-to-stream `JOIN` SQL query in the similar manner as usual JOIN. Let's define `orders`
+and `deliveries` stream events (note, `Stream` type does not exist, used just for example) :
+
+```sql
+CREATE MAPPING orders (order_id BIGINT, order_time TIMESTAMP, item_id BIGINT) TYPE Stream
+CREATE MAPPING deliveries (delivery_id BIGINT, order_id BIGINT, delivery_time TIMESTAMP, address VARCHAR) TYPE Stream
+```
+
+Consider the following query:
+
+```sql
+ CREATE VIEW fast_delivered_orders AS
+ SELECT * FROM IMPOSE_ORDER(TABLE(orders), DESCRIPTOR(order_time), INTERVAL '1' MINUTE) AS o
+ JOIN SELECT * FROM IMPOSE_ORDER(TABLE(deliveries), DESCRIPTOR(order_time), INTERVAL '1' MINUTE) AS d
+ ON d.delivery_time BETWEEN o.order_time AND o.order_time + INTERVAL '1' HOUR
+ 
+> SELECT * FROM fast_delivered_orders
+```
+
+Expected result set is represented in Table 1:
+
+| order_id | order_time       | item_id | delivery_id | order_id | delivery_time    | address     |
+|----------|------------------|---------|-------------|----------|------------------|-------------|
+| 1        | 01.03.2022:10:00 | 100     | 1           | 1        | 01.03.2022:10:40 | 'address'   |
+| 2        | 01.03.2022:11:00 | 100     | 2           | 2        | 01.03.2022:11:20 | 'address_2' |
+| ---      | ---              | ---     | ---         | ---      | ---              | ---         |
+
+__Table 1__
+
+### Technical Design
+
+_This is the technical portion of the design document. Explain the design in sufficient detail._
+
+SQL engine should use a specialized Jet processor to perform JOIN operation for two input stream events. JOIN condition
+allows suppose to be time-bounded. Non-time-bounded JOIN condition is not allowed to be used.
+
+#### Overall design
+
+Consider having two streams - __S1__ and __S2__. There is only one payload in both streams event: conditional timestamp
+of that event.
+
+Let's describe all possible scenarios for joining these two streams:
+
+| Behavior                                                   | Outcome                           |
+|------------------------------------------------------------|-----------------------------------|
+| Both input streams don't have watermarks                   | exception during query processing |
+| At least one input stream doesn't have watermarks          | exception during query processing |
+| At least one input stream has zero lag                     | result set would be empty         |
+| Both __S1__ and __S2__ have watermarks with same lag times | Example in Table 3,4              |
```

Comment:
What if the lag times are different?

Reply:
Yeap, I forgot to mention that. We the only thing that has influence on storing events in buffers is JOIN condition. 
My understanding is that no superior changes are expected with different lag times, but I may be wrong.